syntax = "proto3";

import "empty.proto";

package kaonic;

enum TrxType {
    RF09 = 0;
    RF24 = 1;
 }

service Radio {
    rpc Configure(ConfigurationRequest) returns (kaonic.Empty) {}
    rpc Transmit(TransmitRequest) returns (TransmitResponse) {}
    rpc Receive(ReceiveRequest) returns (ReceiveResponse) {}
    rpc ReceiveStream(kaonic.Empty) returns (stream RadioFrame) {}
}

message RadioFrame {
    bytes data = 1;
    uint32 length = 2;
    uint32 crc32 = 3;
}

message ConfigurationRequest {
    int32 module = 1;
    TrxType trx_type = 2;
    uint32 freq = 3;
    uint32 channel = 4;
    uint32 channelSpacing = 5;
}

message TransmitRequest {
    int32 module = 1;
    TrxType trx_type = 2;
    RadioFrame frame = 3;
}

message TransmitResponse {
    uint32 latency = 1;
}

message ReceiveRequest {
    int32 module = 1;
    TrxType trx_type = 2;
    uint32 timeout = 3;
}

message ReceiveResponse {
    RadioFrame frame = 1;
    int32 rssi = 2;
    uint32 latency = 3;
}
